### ISSUE TITLE
CustomerService: Call to a member function generateHttpHeaders() on null

### ISSUE THREAD

**Issue by minasm** [googleads/googleads-php-lib#221]
#220
@fiboknacky thanks again for your reply and your help! :)

Please note, I cannot close or comment other issues so needed to open two new ones. 
Github keeps telling me;

> You can't comment at this time.

This is my code now and when I run it and get a var_dump on $customerService I see clientCustomerId in the object but it is null.

```
//Setting 3 vars
$clientId = "******";
$clientSecret = "*****";
$developerToken = "*****";
//Redirected here from auth.php with a token.
if(isset($_SESSION['refreshToken'])){
     $refreshToken =  $_SESSION['refreshToken'];
}
$oAuth2Credential = (new OAuth2TokenBuilder())
    ->withClientId($clientId)
    ->withClientSecret($clientSecret)
    ->withRefreshToken($refreshToken)
    ->build();
$session = (new AdWordsSessionBuilder())
    ->withDeveloperToken($developerToken)
    ->withOAuth2Credential($oAuth2Credential)
    ->build();
$adWordsServices = new AdWordsServices();
$customerService = $adWordsServices->get(
    $session, CustomerService::class);

```

After this code block, I am planning to shoot CampaignService service, which uses session again and because I could not retrieve ClientCustomerId page returns

> '[AuthenticationError.CLIENT_CUSTOMER_ID_IS_REQUIRED @ ; trigger:'<null>']'

`$campaignService = $adWordsServices->get($session, CampaignService::class);`

**Comment by fiboknacky** [comment]
Hi @minasm 

Could you please try refreshing your browser next time in case it helps the issue that GitHub doesn't allow you to comment (based on this [reported issue](https://github.com/gitlabhq/gitlabhq/issues/4181))? 

As for your original question, the steps would be:
* Create an OAuth2 credential and use it to create `AdWordsSession` (`$session` in your code snippet)
* Use `$session` to create a `CustomerService`: `$customerService = $adWordsServices->get($session, CustomerService::class);`
* Use `$customerService->getCustomers()` (see this [ref](https://developers.google.com/adwords/api/docs/reference/v201609/CustomerService#getcustomers)) to get your customer info, which contains a client customer ID.
* Use this client customer ID to create a new `AdWordsSession` using [`withClientCustomerId()`](https://github.com/googleads/googleads-php-lib/blob/master/src/Google/AdsApi/AdWords/AdWordsSessionBuilder.php#L209) method.
* Use this new session to call `CampaignService.get()`.

In other words, you can't call the `CampaignService.get()` directly as now it requires the `clientCustomerId` populated in the header, which in turn is fetched from an `AdWordsSession`.
That's why when you call the `CampaignService` directly, the error was thrown.

Hope this help.
Let me know if you have further questions. :)

Best,
Knack

**Comment by komirad** [comment]
About 

> You can't comment at this time.

I faced a similar issue till I disabled ad blockers on github.

**Comment by fiboknacky** [comment]
Thanks for the tip. :)

**Comment by minasm** [comment]
Thanks for the tips! worked for me.
@fiboknacky thanks for all the help! I might ask one more question actually.
I am trying to get the keyword quality score. In old library, it was easier:P
When I try to reach it with this method
`getQualityInfo()`
I am retrieving the data 

`  0 => 
    array (size=4)
      'keywordId' => int 39330000
      'keywordText' => string 'KeywordsText' (length=17)
      'keywordMatchType' => string 'EXACT' (length=5)
      'qualityScore' => 
        object(Google\AdsApi\AdWords\v201609\cm\QualityInfo)[55]
          protected 'qualityScore' => int 4`

So all good, but I want to reach the `qualityScore` and get it into an array. How can I reach it? I found a `getQualityScore()` but it is not helping at all.

Thanks for your advice!

**Comment by fiboknacky** [comment]
Hello @minasm 

If this doesn't bother you too much, could you please open a new issue instead?
This is so that we can keep track of all independent issues and help other users find your question more easily. :-) 
I'll close this issue soon as it seems you have no more questions about `CustomerService` now, but feel free to reopen it if you do have.

Thanks in advance.
Knack

**Comment by kidker** [comment]
[MY FINAL CODE](https://gist.github.com/kidker/ed5c4bc5da20614ee09ed0b96cd741c8)

**Comment by fiboknacky** [comment]
Thanks!
